build-scripts/collect_remediations.py

```diff
@@ -0,0 +1,117 @@
+
+import sys
+import os
+import os.path
+import argparse
+
+import ssg.build_remediations as remediation
+import ssg.build_yaml
+import ssg.rules
+import ssg.jinja
+import ssg.environment
+import ssg.utils
+import ssg.xml
+
+
+def parse_args():
+    p = argparse.ArgumentParser()
+    p.add_argument(
+        "--build-config-yaml", required=True,
+        help="YAML file with information about the build configuration. "
+        "e.g.: ~/scap-security-guide/build/build_config.yml"
+    )
+    p.add_argument(
+        "--product-yaml", required=True,
+        help="YAML file with information about the product we are building. "
+        "e.g.: ~/scap-security-guide/rhel7/product.yml"
+    )
+    p.add_argument(
+        "--resolved-rules-dir", required=True,
+        help="Directory with <rule-id>.yml resolved rule YAMLs"
+    )
+    p.add_argument(
+        "--remediation-type", required=True, action="append",
+        help="language or type of the remediations we are combining."
+        "example: ansible")
+    p.add_argument(
+        "--output-dir", required=True,
+        help="output directory where all remediations will be saved"
+    )
+    p.add_argument(
+        "--fixes-from-templates-dir", required=True,
+        help="directory from which we will collect fixes generated from "
+        "templates")
+
+    return p.parse_args()
+
+
+def prepare_output_dirs(output_dir, remediation_types):
+    output_dirs = dict()
+    for lang in remediation_types:
+        language_output_dir = os.path.join(output_dir, lang)
+        if not os.path.exists(language_output_dir):
+            os.makedirs(language_output_dir)
+        output_dirs[lang] = language_output_dir
+    return output_dirs
+
+
+def combine_remediations(
+        rule, langs, fixes_from_templates_dirs, product, output_dirs,
+        env_yaml):
+    for lang in langs:
+        rule_dir = os.path.dirname(rule.definition_location)
+        ext = remediation.REMEDIATION_TO_EXT_MAP[lang]
+        remediation_cls = remediation.REMEDIATION_TO_CLASS[lang]
+        language_fixes_from_templates_dir = os.path.join(
+            fixes_from_templates_dirs, lang)
+        fix_path = None
+        # first look for a static remediation
+        rule_dir_remediations = remediation.get_rule_dir_remediations(
+            rule_dir, lang, product)
+        if len(rule_dir_remediations) > 0:
+            # first item in the list has the highest priority
+            fix_path = rule_dir_remediations[0]
+        if fix_path is None:
+            # check if we have a templated remediation instead
+            if os.path.isdir(language_fixes_from_templates_dir):
+                templated_fix_path = os.path.join(
+                    language_fixes_from_templates_dir, rule.id_ + ext)
+                if os.path.exists(templated_fix_path):
+                    fix_path = templated_fix_path
+        if fix_path is None:
+            # neither static nor templated remediation found
+            continue
```

Comment:
I propose to extract this into a function that returns the "best guess" fix path.
The interesting thing is that if a rule has a "manual" remediation, it is used in favor of the template remediation. I wonder whether this is a good approach - one can disable template content easily in the template invocation, so I would say that template should have precedence, or there could be a warning that there is ambiguous content available.

Reply:
But disabling a language in the `rule.yml` disables it for all products.

If there is a static remediation it may be because the template doesn't suit a specific product and (probably) cannot be easily incorporated into the template.